pubsub/azure/eventhubs/eventhubs.go

```diff
@@ -98,62 +117,301 @@ func subscribeHandler(ctx context.Context, topic string, e *eventhub.Event, hand
 func NewAzureEventHubs(logger logger.Logger) *AzureEventHubs {
 	return &AzureEventHubs{logger: logger}
 }
 
-func parseEventHubsMetadata(meta pubsub.Metadata) (azureEventHubsMetadata, error) {
+func parseEventHubsMetadata(meta pubsub.Metadata) (*azureEventHubsMetadata, error) {
+	b, err := json.Marshal(meta.Properties)
+	if err != nil {
+		return nil, err
+	}
+
 	m := azureEventHubsMetadata{}
+	err = json.Unmarshal(b, &m)
+	if err != nil {
+		return nil, err
+	}
 
-	if val, ok := meta.Properties[connectionString]; ok && val != "" {
-		m.connectionString = val
-	} else {
-		return m, errors.New(missingConnectionStringErrorMsg)
+	if m.ConnectionString == "" && m.EventHubNamespace == "" {
+		return &m, errors.New(missingConnectionStringNamespaceErrorMsg)
 	}
 
-	if val, ok := meta.Properties[storageAccountName]; ok && val != "" {
-		m.storageAccountName = val
-	} else {
-		return m, errors.New(missingStorageAccountNameErrorMsg)
+	if m.ConnectionString != "" && m.EventHubNamespace != "" {
+		return &m, errors.New(bothConnectionStringNamespaceErrorMsg)
 	}
 
-	if val, ok := meta.Properties[storageAccountKey]; ok && val != "" {
-		m.storageAccountKey = val
-	} else {
-		return m, errors.New(missingStorageAccountKeyErrorMsg)
+	return &m, nil
+}
+
+func validateAndGetHubName(connectionString string) (string, error) {
+	parsed, err := conn.ParsedConnectionFromStr(connectionString)
+	if err != nil {
+		return "", err
 	}
+	return parsed.HubName, nil
+}
 
-	if val, ok := meta.Properties[storageContainerName]; ok && val != "" {
-		m.storageContainerName = val
-	} else {
-		return m, errors.New(missingStorageContainerNameErrorMsg)
+func (aeh *AzureEventHubs) ensureEventHub(hubName string) error {
+	if aeh.hubManager == nil {
+		aeh.logger.Errorf("hubManager client not initialized properly.")
+		return fmt.Errorf("hubManager client not initialized properly")
+	}
+	entity, err := aeh.getHubEntity(hubName)
+	if err != nil {
+		return err
+	}
+	if entity == nil {
+		if err := aeh.createHubEntity(hubName); err != nil {
+			return err
+		}
 	}
+	return nil
+}
+
+func (aeh *AzureEventHubs) ensureSubscription(hubName string) error {
+	err := aeh.ensureEventHub(hubName)
+	if err != nil {
+		return err
+	}
+	_, err = aeh.getConsumerGroupsClient()
+	if err != nil {
+		return err
+	}
+	return aeh.createConsumerGroup(hubName)
+}
+
+func (aeh *AzureEventHubs) getConsumerGroupsClient() (*mgmt.ConsumerGroupsClient, error) {
+	if aeh.cgClient != nil {
+		return aeh.cgClient, nil
+	}
+	client := mgmt.NewConsumerGroupsClientWithBaseURI(aeh.managementSettings.AzureEnvironment.ResourceManagerEndpoint,
+		aeh.metadata.SubscriptionID)
+	a, err := aeh.managementSettings.GetAuthorizer()
+	if err != nil {
+		return nil, err
+	}
+	client.Authorizer = a
+	aeh.cgClient = &client
+	return aeh.cgClient, nil
+}
+
+func (aeh *AzureEventHubs) createConsumerGroup(hubName string) error {
+	create, err := aeh.shouldCreateConsumerGroup(hubName)
```

Comment:
Please, add a retry with exponential backoff in all cases where there can be a race condition between multiple sidecar instances. Have a Max retry of 5 or something. This is helpful because users prefer this over relying only on K8s restarts.

Reply:
done